man/man1/nemoinp.1
.TH NEMOINP 1NEMO "11 October 2023"

.SH "NAME"
nemoinp \- parse a user supplied number string

.SH "SYNOPSIS"
.PP
\fBnemoinp [parameter=value] ...

.SH "DESCRIPTION"
\fInemoinp\fP parses an expression supplied by the user, using the
rules as given by \fInemoinp(3NEMO)\fP. The output format can be
defined too, separators, newlines etc. can be choosen at will.
See also the EXAMPLES below.
.PP
It may come in handy to set an array in a shell variable or to
initialize tables, e.g. \fBnemoinp 1:10:2\fP will then
result in \fB1 3 5 7 9\fP.  Although such implied
loops are a common feature of scripting environments, there
are subtle differences:
.nf

.ta +2i +1i
nemoinp		1:10:2
matlab		1:2:10
seq		1 2 10
numpy.arange	1,10,2	last element (10) is never used
IDL		findgen(5)*2+1  -or-  indgen(5)*2+1

.fi

.SH "PARAMETERS"
The following parameters are recognized in any order if the keyword is also
given:
.TP 20
\fBexpression=\fIexpr-string\fP
expression to be parsed [default: empty ].
.TP
\fBsepar=\fIstring\fP
The string which is appended to the format string described in previous
keyword and is used to separate the numbers output. When no string is
supplied, a blank is used.
[default: empty].
.TP
\fBformat=\fIstring\fP
Valid C-format descriptor used in \fIprintf(3)\fP to output
the numbers. When %d is used, integer math is forced.
See also next keyword for output number separator.
[default: \fB%g\fP].
.TP
\fBnewline=t|f\fP
Should a newline be added after each number parsed. This is often
useful in cases \fInemoinp\fP is used to generate the independant
coordinate when tables have to be generated from scratch.
[Default: \fBt\fP].
.TP
\fBnmax=\fIinteger\fP
Size of scratch array to be allocated dynamically for the parsed
number
[default: \fB32768\fP].
.TP
\fBtab=\fItable_file\fP
Optional table output. Default: none
.TP
\fBseed=\fIinitial_seed\fP
Seed for \fIxrandom(3NEMO)\fP. Default: 0
.TP
\fBatof=\fP
Merely provided to test what \fInatof\fP returns. Note that this
function does not perform any error checking and can return the wrong
number. 
.TP
\fBdms=t|f\fP
Force parsing the (optionally comma separated) string(s) as D:M:S.S, converting 
them to fractional degrees. For H:M:S parsing the user is
responsible for the extra factor of 15.
Default: \fBf\fP.

.SH "CAVEATS"
The number of elements in an array is limited, currently hardcoded to 32678 in
\fInemoinp(3NEMO)\fP. If possible, use \fIseq(1)\fP or \fItabgen(1NEMO)\fP. Here
is an example creating a list of numbers 1 .. 10 with these three modes.
The \fBseq\fP method is the fastest.
.EX
      nemoinp 1:10
      seq 10
      tabgen - 10 1 mode=4
.EE

.SH "EXAMPLES"
.nf
   nemoinp 4.2+3.1           simple math
   nemoinp 'sqrt(2)'         math functions
   nemoinp pi+3              some support for funny numbers
   nemoinp 2:20:4            from 2 to 20 in steps of 4 (NEMO command)
   seq 2 4 20                from 2 to 20 in steps of 4 (UNIX command)
   set n=20                  a (t)C-shell variable
   n=20			     a bash/zsh variable
   nemoinp 2:${n}:4          as before, now using a shell variable
   nemoinp 2::4              repeat the value 4 times
   nemoinp '10**[0:3:1]'     log spaced values  1 10 100 1000
   nemoinp 'iflt(1,2,3,4)'   print 3, not 4, because 1 is less than 2
   nemoinp 30:30:15 dms=t    convert hexasegimal 30:30:15 into decimal 30.5043
   nemoinp c,G,h,p,m         a few good constants
   nemoinp 'p*pi/(3600*180)' 1 AU in m
.fi
things that it cannot do are sometimes captured with little known unix commands. Here
are some examples:
.nf
   echo "ibase=2; obase=8; 101101" | bc           # convert binary to octal
   echo 1.2/4.5 | bc -l                           # floating point math
   z=$((x + y))                                   # sum integers in bash
.fi

.SH "SEE ALSO"
seq(1), tabgen(1NEMO), tabmath(1NEMO), nemoinp(3NEMO), getrange(3NEMO), functions(5NEMO)

.SH "AUTHOR"
Peter Teuben

.SH "UPDATE HISTORY"
.nf
.ta +1.0i +4.0i
18-apr-89	V1.0 created	PJT
20-jun-89	V1.0a small changed - doc updated	PJT
 5-jul-89	V1.1 added nmax parameter	PJT
18-may-92	documentation
16-jun-97	V1.6 %d forces integer math	PJT
28-aug-00	V1.7 documented tab= and seed=	PJT
31-may-01	V1.8 added atof=	PJT
4-mar-03	V1.9 added dms=  	PJT
.fi


src/kernel/misc/nemoinp.c
/*   nemoinp (archaic), nemoinpd, nemoinpf, nemoinpi, nemoinpl, nemoinpb
 *                      nemorinpd, nemorinpf, nemorinpi, nemorinpl, nemorinpb
 *                      natof, natoi
 *
 * NEMOINP:	derivatives of herinp(), more c-user friendly
 *		interface of the same
 *  returns a negative number (the herinp() error code) when a 
 *  parsing error occurred, and a positive (or 0) number of the
 *  the number of elements returned in 'a' when no error detected.
 *
 *  18-May-88   Peter Teuben
 *   6-Dec-88   added 'elen' parameter because of new herinp PJT
 *   6-Mar-89   split into nemoinpX functions
 *   5-jul-89   can't make up my mind To have an _ or not to have.. PJT
 *  14-oct-90   allow option with burststring and using getXparam.. PJT
 *  25-feb-92	happy gcc2.0 - kludged up NOHERINP		    PJT
 *   7-jan-93   bit more help for ignorant user in the TOOLBOX version PJT
 *   4-mar-94   V1.3 ansi - added nemoinpf for float's              pjt
 *   2-jul-95   V1.4 added tab= to have optional table output file  pjt
 *  18-sep-96   V1.5 changed default of newline from 'f' to 't'     pjt
 *  16-jun-97   V1.6 use integer  routines if %d is used	    pjt
 *  13-mar-00   V1.7 added seed= parameter			    pjt
 *                   (but it doesn't appear to do anything useful)
 *  31-may-01   V1.8 added natof, natoi                             pjt
 *   7-jun-01       a:  added casting to shut up compiler           pjt
 *   8-sep-01       b:  init_xrandom
 *   4-mar-03   V1.9: added nemoinpx for sexa decoding into degrees pjt
 *  28-jun-03       a: fixed prototype for darwin :-)               pjt
 *  24-nov-03       b: fixed another prototype for gcc3             pjt
 *  28-jan-04       c: recognize nan or NaN and return same         pjt
 *   6-jan-05   V1.10:  added the repeat nemorinpX routines         pjt
 *   5-may-05       a: add newline                                  pjt
 *   5-feb-06       b: also recognize -nan or -NaN ....             pjt
 *
 * TODO:  how to handle largest and smallest number (old style MIN/MAXLOG 38)
 */

#include <stdinc.h>
#include <getparam.h>
#include <extstring.h>

extern string *burststring(string,string);
extern void freestrings(string *);

#if defined(NOHERINP)

int nemoinpi(
	     char *expr,
	     int  *a,
	     int   na)
{
  int nret;
  string *vals;

    vals = burststring(expr,", ");
    for (nret=0; vals[nret] != NULL; nret++) {
        if (nret>=na) {
            return(-23);
	}
        a[nret] = atoi(vals[nret]);
    }
    return(nret);    
}

int nemoinpl(
	     char *expr,
	     long *a,
	     int   na)
{
  int nret;
  string *vals, *burststring();

    vals = burststring(expr,", ");
    for (nret=0; vals[nret] != NULL; nret++) {
        if (nret>=na) {
            return(-23);
        }
        a[nret] = atoi(vals[nret]);
    }
    return(nret);
}


int nemoinpb(
	     char *expr,
	     bool *a,
	     int   na)
{
  int nret;    
  string *vals;

    vals = burststring(expr,", ");
    for (nret=0; vals[nret] != NULL; nret++) {
        if (nret>=na) {
            return(-23);
	}
        a[nret] = (*vals[nret]=='t' || *vals[nret]=='T') ? TRUE : FALSE;
    }
    return(nret);    
}
    

int nemoinpd(
	     char   *expr,
	     double *a,
	     int     na)
{
  int nret;
  string *vals;

    vals = burststring(expr,", ");
    for (nret=0; vals[nret] != NULL; nret++) {
        if (nret>=na) {
            return(-23);
        }
        a[nret] = atof(vals[nret]);
    }
    return(nret);
}
    

int nemoinpf(
	     char   *expr,
	     float  *a,
	     int     na)
{
  int nret;
  string *vals;

    vals = burststring(expr,", ");
    for (nret=0; vals[nret] != NULL; nret++) {
        if (nret>=na) {
            return(-23);
        }
        a[nret] = atof(vals[nret]);
    }
    return(nret);
}


#else

extern void herinp(char *expr, int *nchr, char *type, int *length,
		   char *outv, int *nout, int *nret, int *ierd);


int nemoinpi(
	     char *expr,
	     int  *a,
	     int   na)
{
    int ierr, nret, elen, tlen;
    char type;

    elen = strlen(expr);
    tlen = sizeof(int);
    type = 'I';     /* integer */
    herinp (expr, &elen, &type, &tlen, (char *)a, &na, &nret, &ierr);
    if (ierr < 0)
        return(ierr);
    else
        return(nret);
}

int nemoinpl(
	     char *expr,
	     long *a,
	     int   na)
{
    int ierr, nret, elen, tlen;
    char type;

    elen = strlen(expr);
    tlen = sizeof(long);
    type = 'I'; /* integer */
    herinp (expr, &elen, &type, &tlen, (char *)a, &na, &nret, &ierr);
    if (ierr < 0)
        return(ierr);
    else
        return(nret);
}


int nemoinpb(
	     char *expr,
	     bool *a,
	     int   na)
{
    int ierr, nret, elen, tlen;
    char type;

    elen = strlen(expr);
    tlen = sizeof(bool);
    type = 'L';     /* logical */
    herinp (expr, &elen, &type, &tlen, (char *)a, &na, &nret, &ierr);
    if (ierr < 0)
        return(ierr);
    else
        return(nret);
}
    

int nemoinpd(
	     char   *expr,
	     double *a,
	     int     na)
{
    int ierr, nret, elen, tlen;
    char type;

    elen = strlen(expr);
    tlen = sizeof(double);
    type = 'F';
    herinp (expr, &elen, &type, &tlen, (char *)a, &na, &nret, &ierr);
    if (ierr < 0)
        return(ierr);
    else
        return(nret);
}

int nemoinpf(
	     char   *expr,
	     float  *a,
	     int     na)
{
    int ierr, nret, elen, tlen;
    char type;

    elen = strlen(expr);
    tlen = sizeof(float);
    type = 'F';
    herinp (expr, &elen, &type, &tlen, (char *)a, &na, &nret, &ierr);
    if (ierr < 0)
        return ierr;
    else
        return nret;
}
#endif

/*
 *  convert Babylonian sexagesimal string
 *  D:M:S -> D + M/60 + S/3600
 *    @todo   H:M:S is normal,
 *        NED  example 00h07m15.84s, +27d42m29.1s
 *        CASA example 12:22:55.212, +15.49.15.500    (who came up with THAT?)
 *        ???          10,20,30.456                 (miriad might have an input example)
 *
 *    sep options:      ':'  ','   '.'  'hms'   'dms'
 */

double atox(string nptr)
{
  int ncomp, sign;
  string *comp;
  double retval;
  
  comp = burststring(nptr,":");
  ncomp = xstrlen(comp,sizeof(string))-1;
  if (ncomp < 1)
    return 0.0;
  sign = (*comp[0]=='-') ?   -1  : +1;
  retval = fabs(atof(comp[0]));
  if (ncomp > 1) {
    retval += atof(comp[1])/60.0;
    if (ncomp > 2) {
      retval += atof(comp[2])/3600.0;
      if (ncomp > 3)
	warning("String %s has too many : for atox(), ignoring remainder",nptr);
    }
  }
  retval *= sign;
  freestrings(comp);
  return retval;
}

int nemoinpx(
	     char *expr,
	     double *a,
	     int     na)
{
  int nret, ncomp, sign;
  string *vals, *comp;

  vals = burststring(expr,",");
  for (nret=0; vals[nret] != NULL; nret++) {
    if (nret>=na)
      return -23;
#if 1
    // this code is the error checking version of atox()
    // @todo  check for non-digits
    comp = burststring(vals[nret],":");
    ncomp = xstrlen(comp,sizeof(string))-1;
    if (ncomp < 1 || ncomp > 3)
      return -13;
    sign = (*comp[0]=='-') ?   -1  : +1;
    a[nret] = fabs(atof(comp[0]));
    if (ncomp > 1) {
      a[nret] += atof(comp[1])/60.0;
      if (ncomp > 2)
	a[nret] += atof(comp[2])/3600.0;
    }
    a[nret] *= sign;
    freestrings(comp);
#else
    a[nret] = atox(vals[nret]);
#endif
  }
  freestrings(vals);
  return nret;
}


int nemorinpi(string var, int *xvar,  int n, int defvar, bool repeated)
{
  int i, nret;
  if (var == 0 || *var == 0) {
    for (i=0; i<n; i++) 
      xvar[i] = defvar;
    return 0;
  } else {
    nret = nemoinpi(var,xvar,n);
    if (nret <= 0) error("nemorinpi: parsing error %d in %s",nret,var);
    for (i=nret; i<n; i++)
      xvar[i] = repeated ? xvar[i-1] : defvar;
    return nret;
  }
  return -1;
}


int nemorinpd(string var,  double *xvar, int n,double defvar, bool repeated)
{
  int i, nret;
  if (var == 0 || *var == 0) {
    for (i=0; i<n; i++) 
      xvar[i] = defvar;
    return 0;
  } else {
    nret = nemoinpd(var,xvar,n);
    if (nret <= 0) error("nemorinpd: parsing error %d in %s",nret,var);
    for (i=nret; i<n; i++)
      xvar[i] = repeated ? xvar[i-1] : defvar;
    return nret;
  }
  return -1;
}

int nemorinpf(string var, float *xvar,  int n, float defvar, bool repeated)
{
  int i, nret;
  if (var == 0 || *var == 0) {
    for (i=0; i<n; i++) 
      xvar[i] = defvar;
    return 0;
  } else {
    nret = nemoinpf(var,xvar,n);
    if (nret <= 0) error("nemorinpf: parsing error %d in %s",nret,var);
    for (i=nret; i<n; i++)
      xvar[i] = repeated ? xvar[i-1] : defvar;
    return nret;
  }
  return -1;
}

int nemorinpl(string var, long *xvar,  int n, long defvar, bool repeated)
{
  int i, nret;
  if (var == 0 || *var == 0) {
    for (i=0; i<n; i++) 
      xvar[i] = defvar;
    return 0;
  } else {
    nret = nemoinpl(var,xvar,n);
    if (nret <= 0) error("nemorinpl: parsing error %d in %s",nret,var);
    for (i=nret; i<n; i++)
      xvar[i] = repeated ? xvar[i-1] : defvar;
    return nret;
  }
  return -1;
}

int nemorinpb(string var, bool *xvar,  int n, bool defvar, bool repeated)
{
  int i, nret;
  if (var == 0 || *var == 0) {
    for (i=0; i<n; i++) 
      xvar[i] = defvar;
    return 0;
  } else {
    nret = nemoinpb(var,xvar,n);
    if (nret <= 0) error("nemorinpb: parsing error %d in %s",nret,var);
    for (i=nret; i<n; i++)
      xvar[i] = repeated ? xvar[i-1] : defvar;
    return nret;
  }
  return -1;
}





double natof(char *expr)
{
  double x;
  int n;
  if (streq(expr, "nan") || streq(expr, "NaN") || streq(expr,"-nan") || streq(expr,"-NaN"))
    return atof("nan");

  n = nemoinpd(expr,&x,1);
  if (n!=1) warning("natof parsing: %s",expr);
  return x;
}

int natoi(char *expr)
{
  int x, n;
  n = nemoinpi(expr,&x,1);
  if (n!=1) warning("natof parsing: %s",expr);  
  return x;
}


#if defined(TOOLBOX)

string defv[] = {
    "expression=\n	Expression to parse [help]",
    "separ=\n		Separator between numbers on output",
    "format=%g\n	Format used to print numbers",
    "newline=t\n	Use newline between numbers?",
    "nmax=100000\n	Size of output buffer",
    "tab=\n             Optional table output",
    "seed=0\n		Seed for xrandom",
    "atof=\n            test (n)atof single value expression",
    "dms=f\n            Use D:M:S.SS parsing instead of regular",
    "VERSION=1.11\n	11-mar-2022 PJT",
    NULL,
};


string usage = "expression parser and evaluator; also does lists";


void nemo_main(void)
{
    char   fmt1[20], fmt2[20], *cp;
    double dms[32];
    double *x;
    int    *ix;
    int    i,nret, nx, seed;
    bool   Qnl, Qint,Qdms;
    stream outstr;

    if (hasvalue("tab"))
        outstr = stropen(getparam("tab"),"w");
    else
        outstr = stdout;                /* is that really ok? */
    cp = getparam("format");            /* get format string to print with */
    if (strchr(cp,'%')==NULL)
      warning("%s badly formed printf-conversion specification ???",cp);
    if (strchr(cp,'d')) {
      dprintf(1,"Using integer math\n");
      Qint = TRUE;
    } else {
      dprintf(1,"Using floating point math\n");
      Qint = FALSE;
    }
    seed = init_xrandom(getparam("seed"));
    dprintf(1,"init_xrandom: seed=%d\n",seed);

    strcpy (fmt1,cp);                   /* store it in 'fmt' */
    strcpy (fmt2,cp);                   /* and here */
    
    Qdms = getbparam("dms");
    if (Qdms) {
      strcat(fmt1,"\n");
      nret = nemoinpx(getparam("expression"),dms,32);
      if (nret < 0) error("Parsing dms expression");
      for (i=0; i<nret; i++)
	printf(fmt1,dms[i]);
      return;  /* for now */
    }

    cp = getparam("separ");             
    if (hasvalue("separ"))              /* separator between numbers ? */
        strcat(fmt2,getparam("separ"));
    else
        strcat (fmt2," ");              /* else use blank as separator */

    Qnl = getbparam("newline");         /* use newlines also ? */

    if (hasvalue("atof")) {
      printf(fmt1,natof(getparam("atof")));
      printf("\n");
      return;
    }

    if (!hasvalue("expression")) {      /* extra help if nothing given */
    	/* this somewhat unusual exit is because aliens often use this program */
	/* what really should have been done is make "expression=???" default */
        dprintf(0,"Usage: %s <expression>\n",getargv0());
        dprintf(0,"<expr> can be of form:  start[:end][:incr][,start::repeat]...\n");
	dprintf(0,"NEMO=%s\n",getenv("NEMO"));
        stop(0);
    }
    nx = getiparam("nmax");

    if (Qint) {
        ix = (int *) allocate(nx * sizeof(int));
   
        nret = nemoinpi(getparam("expression"),ix,nx);
        if (nret == -23)
            error("Too many items in list, use a bigger nmax=%d",nx);
        else if (nret < 0)
            error("nemoinp (%s) parsing error (%d)",cp,nret);

        for (i=0; i< nret-1; i++) {
            fprintf(outstr,fmt2,ix[i]);
            if (Qnl)
                fprintf(outstr,"\n");
        }
        fprintf (outstr,fmt1,ix[nret-1]);
        if (Qnl)
            fprintf (outstr,"\n");
    } else {
        x = (double *) allocate(nx * sizeof(double));
   
        nret = nemoinpd(getparam("expression"),x,nx);
        if (nret == -23)
            error("Too many items in list, use a bigger nmax=%d",nx);
        else if (nret < 0)
            error("nemoinp (%s) parsing error (%d)",cp,nret);

        for (i=0; i< nret-1; i++) {
            fprintf(outstr,fmt2,x[i]);
            if (Qnl)
                fprintf(outstr,"\n");
        }
        fprintf (outstr,fmt1,x[nret-1]);
        if (Qnl)
            fprintf (outstr,"\n");
    }
    if (!Qnl) fprintf(outstr,"\n");  /* always make sure to add 1 newline */

}
#endif
